src/tools/rbd_mirror/image_map/ImageAction.cc

```diff
@@ -0,0 +1,362 @@
+  if (mirror_uuid.empty()) {
+    return MIRROR_PEER_LOCAL;
+  } else {
+    return MIRROR_PEER_REMOTE;
+  }
+}
+
+ImageAction::ImageAction()
+  : m_lock(unique_lock_name("rbd::mirror::image_map::ImageAction::m_lock", this)) {
+}
+
+ImageAction::Peer &ImageAction::peer_type(ActionState &action_state, MirrorPeerType type) {
+  switch (type) {
+  case MIRROR_PEER_LOCAL:
+    return action_state.local;
+  case MIRROR_PEER_REMOTE:
+    return action_state.remote;
+  default:
+    derr << ": unknown peer type (" << static_cast<uint32_t>(type) << ")" << dendl;
+    assert(false);
+  }
+}
+
+void ImageAction::init_state(const std::string &global_image_id) {
+  Mutex::Locker locker(m_lock);
+
+  dout(20) << ": global_image_id=" << global_image_id << dendl;
+
+  auto ins = m_actions.emplace(global_image_id, ActionState());
+  assert(ins.second);
+}
+
+void ImageAction::release_state(const std::string &global_image_id) {
+  Mutex::Locker locker(m_lock);
+
+  dout(20) << ": global_image_id=" << global_image_id << dendl;
+
+  auto it = m_actions.find(global_image_id);
+  assert(it != m_actions.end());
+
+  m_actions.erase(global_image_id);
+}
+
+bool ImageAction::queue_action(const std::string &global_image_id,
+                               const Action &action, MirrorPeerType type) {
+  Mutex::Locker locker(m_lock);
+
+  dout(20) << ": global_image_id=" << global_image_id << ", action="
+           << action << ", type=" << type << dendl;
+
+  auto it = m_actions.find(global_image_id);
+  assert(it != m_actions.end());
+
+  Peer &peer = peer_type(it->second, type);
+
+  bool can_schedule = peer.actions.empty();
+  peer.actions.push_back(action);
+
+  return can_schedule;
+}
+
+ImageAction::State ImageAction::current_state(const Peer &peer) {
+  assert(m_lock.is_locked());
+
+  return peer.states.first;
+}
+
+void ImageAction::set_next_state(Peer &peer, State state) {
+  assert(m_lock.is_locked());
+
+  dout(20) << ": next_state=" << state << dendl;
+  peer.states.second = state;
+}
+
+ImageAction::State ImageAction::reach_state(Peer &peer) {
+  assert(m_lock.is_locked());
+
+  dout(20) << ": old_state=" << peer.states.first << ", new_state="
+           << peer.states.second << dendl;
+  if (peer.states.first != peer.states.second) {
+    peer.states.first = peer.states.second;
+  }
+
+  return peer.states.first;
+}
+
+// use boost::statechart??
+const ImageAction::Transition
+ImageAction::transition_table[ActionType::ACTION_MAX][STATE_MAX] = {
+  // ACTION_ADD
+  {
+    {SUB_ACTION_MAP,      STATE_ASSOCIATED},    // STATE_UNASSIGNED
+    {SUB_ACTION_COMPLETE, STATE_ASSOCIATED},    // STATE_ASSOCAITED
+    {SUB_ACTION_INVALID,  STATE_UNASSIGNED},    // STATE_DISASSOCIATED
+    {SUB_ACTION_INVALID,  STATE_UNASSIGNED},    // STATE_PURGING
+    {SUB_ACTION_INVALID,  STATE_UNASSIGNED}     // STATE_SHUFFLING
+  },
+  // ACTION_REMOVE
+  {
+    {SUB_ACTION_UNMAP,    STATE_DISASSOCIATED}, // STATE_UNASSIGNED
```

Comment:
The map could still be present on-disk. We should jump to `PURGING` state in that case. I do not clearly recall why I followed the normal removal state transition when the state is `UNASSIGNED`.